spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubMessageSource.java

```diff
@@ -0,0 +1,125 @@
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.pubsub.integration.inbound;
+
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.cloud.gcp.pubsub.core.subscriber.PubSubSubscriberOperations;
+import org.springframework.cloud.gcp.pubsub.integration.AckMode;
+import org.springframework.cloud.gcp.pubsub.integration.PubSubHeaderMapper;
+import org.springframework.cloud.gcp.pubsub.support.GcpPubSubHeaders;
+import org.springframework.cloud.gcp.pubsub.support.converter.ConvertedAcknowledgeablePubsubMessage;
+import org.springframework.integration.endpoint.AbstractMessageSource;
+import org.springframework.integration.mapping.HeaderMapper;
+import org.springframework.util.Assert;
+
+/**
+ * A <a href="https://cloud.google.com/pubsub/docs/pull#pubsub-pull-messages-sync-java">PubSub Synchronous pull</a>
+ * implementation of {@link AbstractMessageSource}.
+ *
+ * Enables more even load balancing with polling adapters, at the cost of efficiency gained from batched pulls.
+ *
+ * @author Elena Felder
+ *
+ * @since 1.1
+ */
+public class PubSubMessageSource extends AbstractMessageSource<Object> {
+
+	private static final Log LOGGER = LogFactory.getLog(PubSubMessageSource.class);
+
+	private final String subscriptionName;
+
+	private final PubSubSubscriberOperations pubSubSubscriberOperations;
+
+	private AckMode ackMode = AckMode.MANUAL;
+
+	private HeaderMapper<Map<String, String>> headerMapper = new PubSubHeaderMapper();
+
+	private Class payloadType = byte[].class;
+
+	public PubSubMessageSource(PubSubSubscriberOperations pubSubSubscriberOperations, String subscriptionName) {
+		Assert.notNull(pubSubSubscriberOperations, "Pub/Sub subscriber template can't be null.");
+		Assert.notNull(subscriptionName, "Pub/Sub subscription name can't be null.");
+		this.pubSubSubscriberOperations = pubSubSubscriberOperations;
+		this.subscriptionName = subscriptionName;
+	}
+
+	public AckMode getAckMode() {
+		return this.ackMode;
+	}
+
+	public void setAckMode(AckMode ackMode) {
+		this.ackMode = ackMode;
+	}
+
+	public Class getPayloadType() {
+		return this.payloadType;
+	}
+
+	public void setPayloadType(Class payloadType) {
+		this.payloadType = payloadType;
+	}
+
+	public HeaderMapper<Map<String, String>> getHeaderMapper() {
+		return this.headerMapper;
+	}
+
+	public void setHeaderMapper(HeaderMapper<Map<String, String>> headerMapper) {
+		this.headerMapper = headerMapper;
+	}
+
+	@Override
+	protected Object doReceive() {
+		List<ConvertedAcknowledgeablePubsubMessage> messages
```

Comment:
You probably need a try-catch around the whole method and nack, if `AUTO` ack mode is on.

Reply:
I am not sure that's useful, since it does not tell us anything about what happened during message processing. A channel adapter, like PubSubInboundChannelAdapter, handles sending to output, so it can observe whether downstream processing failed. A message source like PubSubMessageSource merely provides the source data; it does not have visibility into what happens at destination. That's a concern of SourcePollingChannelAdapter, which we don't control.

I think users of pollable message source will just have to opt for manual acking if "ack when the message reached Spring Integration infrastructure" is not sufficient.

Reply:
You can take a look what we don in the `AmqpMessageSource`: https://github.com/spring-projects/spring-integration/blob/master/spring-integration-amqp/src/main/java/org/springframework/integration/amqp/inbound/AmqpMessageSource.java
or `KafkaMessageSource`: https://github.com/spring-projects/spring-integration-kafka/blob/master/src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageSource.java

Reply:
Ah, so the class would add its own callback header. I'll do that, thank you!